visualization/app/codeCharta/state/store/metricData/edgeMetricData/edgeMetricData.service.ts

```diff
@@ -0,0 +1,111 @@
+import { StoreService, StoreSubscriber } from "../../../store.service"
+import { IRootScopeService } from "angular"
+import { calculateNewEdgeMetricData, EdgeMetricDataActions } from "./edgeMetricData.actions"
+import { isActionOfType } from "../../../../util/reduxHelper"
+import {
+	AttributeTypes,
+	AttributeTypeValue,
+	BlacklistItem,
+	CodeMapNode,
+	EdgeMetricCount,
+	EdgeMetricData
+} from "../../../../codeCharta.model"
+import { FileState } from "../../../../model/files/files"
+import { HierarchyNode } from "d3"
+import { BlacklistService, BlacklistSubscriber } from "../../fileSettings/blacklist/blacklist.service"
+import { FilesSelectionSubscriber, FilesService } from "../../files/files.service"
+import { nodeEdgeMetricsMap } from "./edgeMetricData.reducer"
+import { AttributeTypesService, AttributeTypesSubscriber } from "../../fileSettings/attributeTypes/attributeTypes.service"
+
+export interface EdgeMetricDataSubscriber {
+	onEdgeMetricDataChanged(edgeMetricData: EdgeMetricData[])
+}
+
+export class EdgeMetricDataService implements StoreSubscriber, BlacklistSubscriber, FilesSelectionSubscriber, AttributeTypesSubscriber {
+	private static EDGE_METRIC_DATA_CHANGED_EVENT = "edge-metric-data-changed"
+	public static NONE_METRIC = "None"
+
+	constructor(private $rootScope: IRootScopeService, private storeService: StoreService) {
+		StoreService.subscribe(this.$rootScope, this)
+		BlacklistService.subscribe(this.$rootScope, this)
+		FilesService.subscribe(this.$rootScope, this)
+		AttributeTypesService.subscribe(this.$rootScope, this)
+	}
+
+	public onStoreChanged(actionType: string) {
+		if (isActionOfType(actionType, EdgeMetricDataActions)) {
+			this.notify(this.select())
+		}
+	}
+
+	public onBlacklistChanged(blacklist: BlacklistItem[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(this.storeService.getState().files, blacklist))
+	}
+
+	public onFilesSelectionChanged(files: FileState[]) {
+		this.storeService.dispatch(calculateNewEdgeMetricData(files, this.storeService.getState().fileSettings.blacklist))
+	}
+
+	public onAttributeTypesChanged(attributeTypes: AttributeTypes) {
+		this.storeService.dispatch(
+			calculateNewEdgeMetricData(this.storeService.getState().files, this.storeService.getState().fileSettings.blacklist)
+		)
+	}
+
+	public getMetricNames(): string[] {
+		if (!this.storeService.getState().metricData.edgeMetricData) {
+			return []
+		}
+
+		return this.storeService.getState().metricData.edgeMetricData.map(x => x.name)
+	}
+
+	public getAmountOfAffectedBuildings(metricName: string): number {
+		if (!nodeEdgeMetricsMap || !nodeEdgeMetricsMap.has(metricName)) {
+			return 0
+		}
+		return nodeEdgeMetricsMap.get(metricName).size
```

Comment:
```suggestion
		const nodeEdgeMetrics = nodeEdgeMetricsMap.get(metricName)
		return nodeEdgeMetrics === undefined ? 0 : nodeEdgeMetrics.size
```

The map should always be initialized when this is called. I would also not use `has` and `get` together. Instead, always use `get` and check for `undefined`. That way there's one less lookup necessary.